Lombiq.Tests.UI/SecurityScanning/SecurityScanningUITestContextExtensions.cs

```diff
@@ -107,6 +116,11 @@ public static class SecurityScanningUITestContextExtensions
         Action<SecurityScanConfiguration> configure = null,
         Action<SarifLog> assertSecurityScanResult = null)
     {
+        // Verify that the app logs are fine right now, then disable app log assertion for the duration of this scan.
+        await context.Configuration.AssertAppLogsAsync(context.Application);
+        var assertAppLogsAsync = context.Configuration.AssertAppLogsAsync;
+        context.Configuration.AssertAppLogsAsync = _ => Task.CompletedTask;
```

Comment:
As it was, if anything breaks the app log assertion kicks you out before the ZAP report is finished. So we actually never get any useful information about correct error handling. Even though a lot of security vulnerabilities are related to (incorrect) error handling.
![image](https://github.com/Lombiq/UI-Testing-Toolbox/assets/4007293/2a92fce0-75fa-44d0-80ab-34ecd251e620)

Reply:
I see that as a smaller issue than not having app log assertions at all.

So, if I understand correctly, currently you get an exception and test failed if ZAP causes log entries, but you won't get the ZAP report. So, you need to fix the logged issues before you can progress with fixing whatever else ZAP may discover (ZAP shouldn't cause exceptions in your app, only input validation errors at most, but those needn't be logged). I don't see how this is survivorship bias (in the end, you need to fix both the logged errors and the ZAP alerts, it's just that won't see both at the same time). With this change, you won't see the error details from the logs at all (keep in mind that some errors are logged but aren't surfaced on the UI).

It would be nice if you'd still get the ZAP report, but then that needs to be fixed, not disabling log assertions for everything that ZAP does.

Reply:
I still think it's a big deal, but I don't have the energy to debate this (I'm sick). Anyway, [we can have an app log assertion at the end](https://github.com/Lombiq/UI-Testing-Toolbox/blob/6fa5d43849c1b8bef9489437b611ad55e352ab87/Lombiq.Tests.UI/SecurityScanning/SecurityScanningUITestContextExtensions.cs#L138). This way the logs are still verified if the scan passes. If the scan doesn't pass, you are supposed to look through the log file in the failure dump anyway.

Reply:
I don't want to debate for the debate's sake, but I honestly don't understand what benefits your original changes bring.

Now what's there is that the security scan failing has a higher priority than the log assertion failing, right? So, if the scan fails and there are log entries, then the test will fail with a `SecurityScanningAssertionException` instead of a Shouldy exception (coming from the log assertion). Once you fix the issues found the scan, only then will the test fail on the log assertion, if any log entries remain.

I see the appeal of having the scan first, but I'm afraid that this can lead to people configuring the spider to not visit certain pages, causing no errors to be logged.

Get well!

Reply:
> Now what's there is that the security scan failing has a higher priority than the log assertion failing, right?

Correct.

> I'm afraid that this can lead to people configuring the spider to not visit certain pages, causing no errors to be logged.

At least for my first use the opposite was true. I made this change initially, to avoid having to blacklist error pages when I just wanted to see the zap report already.

> Get well!

Thanks!

Reply:
What kind of error pages? How did the ZAP report help with fixing the errors?

Reply:
I mean when ZAP does weird stuff (e.g. fuzzing) that causes errors when normal human behavior doesn't. It makes things harder if I can't see what ZAP was doing (e.g. the request it sent is in the report). This has already happened during this issue, so I speak from real life experience when I say that I found the current exception priority better.

Reply:
I see the DX problem, but I don't see it in the code. Looking into it, in the current `dev` code, an assertion exception will be thrown [here](https://github.com/Lombiq/UI-Testing-Toolbox/blob/2aca656f7ae9cd2478a8a93f337619c07b7b6d70/Lombiq.Tests.UI/SecurityScanning/SecurityScanningUITestContextExtensions.cs#L118) if the scan fails. This happens before the logs are asserted [here](https://github.com/Lombiq/UI-Testing-Toolbox/blob/2aca656f7ae9cd2478a8a93f337619c07b7b6d70/Lombiq.Tests.UI/Services/UITestExecutionSession.cs#L116). So if both the scan and the log assertion would fail, then even in the `dev` code, the test will fail with a `SecurityScanningAssertionException`, and both the ZAP report and the logs will be in the failure dump. I just tested this.

What can happen though is that the ZAP scan actually doesn't fail, but still, it causes exceptions in the background, which causes the log assertion to fail. And in that case, it'd be useful to have the ZAP report.

There's actually [a config to still generate the ZAP report](https://github.com/Lombiq/UI-Testing-Toolbox/blob/2aca656f7ae9cd2478a8a93f337619c07b7b6d70/Lombiq.Tests.UI/SecurityScanning/SecurityScanningConfiguration.cs#L20), but I didn't think of such a use case, and for the original intended one (to generate a report even if the test passes) it's broken. But for this it's useful, so you can make it public and enable it during the test in question (or during debugging).

Reply:
I get `PageChangeAssertionException` while the security scan is running. It's not caused by the assertion at the end of the test that you linked.

Reply:
Where do you get the `PageChangeAssertionException` from, exactly? I don't see how that would happen in the current `dev`, since there's no page change between the scan starting and its assertion.

Reply:
I can no longer reproduce it, so for now I have deleted this log suppression. (actually I've converted it into an extension method because I found another use for it, more below)